### PR TITLE
feat: define ast

### LINKED ISSUE #14: Design AST
Design AST

reference [Filter database entries](https://developers.notion.com/reference/post-database-query-filter)

### PR DESCRIPTION
**Pull request by MrHeer**
resolve #14

### CODE REVIEW

File: packages/core/src/ast.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      type?: "created_time";
+    }
+  | {
+      last_edited_time: DatePropertyFilter;
+      property: string;
+      type?: "last_edited_time";
+    }
+  | { formula: FormulaPropertyFilter; property: string; type?: "formula" };
+
+export type TimestampCreatedTimeFilter = {
```
Comment (thomaszdxsn) [inline packages/core/src/ast.ts:100]: remove this

File: packages/core/src/ast.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  created_time: DatePropertyFilter;
+  timestamp: "created_time";
+  type?: "created_time";
+};
+
+export type TimestampLastEditedTimeFilter = {
```
Comment (thomaszdxsn) [inline packages/core/src/ast.ts:106]: remove this

File: packages/core/src/ast.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  type?: "last_edited_time";
+};
+
+export type Filter =
+  | {
+      or: Array<
+        | PropertyFilter
+        | TimestampCreatedTimeFilter
+        | TimestampLastEditedTimeFilter
+        | { or: Array<PropertyFilter> }
```
Comment (thomaszdxsn) [inline packages/core/src/ast.ts:118]: if interface like this, the max level is 2.

we should define more generic interface

File: packages/core/src/ast.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
+// reference https://github.com/makenotion/notion-sdk-js/blob/ebf5c7411b331b8e4930151d22c539a248ac749c/src/api-endpoints.ts#L9219
+
+export type EmptyObject = Record<string, never>;
+
+export type ExistencePropertyFilter =
+  | { is_empty: true }
```
Comment (thomaszdxsn) [inline packages/core/src/ast.ts:6]: camelCase, please

**Review by thomaszdxsn — CHANGES_REQUESTED** [review]

**Review by ZZITE — APPROVED** [review]

**Review by thomaszdxsn — APPROVED** [review]

**Pull request merged**